Implement Ingredient models

sc-arecrow opened the pull request:
Add ingredient models

# How to use
* Refer to documentation within

# TODO
* [x] Integrate models with existing storage models
* [x] Update current storage tests
* [x] Finish writing tests
* [ ] Integrate category changes to database
* [x] Redesign `Quantity` into a class

sc-arecrow committed:
Add DS store to gitignore

sc-arecrow committed:
Refactor ingredient sets to batches

sc-arecrow committed:
Integrate ingredient models with database

sc-arecrow committed:
Add tests for Quantity and IngredientCategory

chrystalquek commented on ChopChop/Models/Ingredients/Ingredient.swift:
```diff
@@ -0,0 +1,271 @@
+ - Each batch has a unique expiry date.
+ */
+class Ingredient: FetchableRecord {
+    var id: Int64?
```
Actually is there a reason why it is a variable and optional?

sc-arecrow replied:
Making the `id`s `private(set) var` would probably be better.
@seowalex Can the `id` be non-optional?

seowalex replied:
It's optional because when you first create an Ingredient, it doesn't have an `id`. It'll only be populated by the save methods after saving, which it's why it's variable.

chrystalquek commented on ChopChop/Models/Ingredients/IngredientCategory.swift:
```diff
@@ -0,0 +1,59 @@
+        self.id = id
+        self.name = name
+
+        ingredients.forEach { $0.ingredientCategoryId = id }
```
it will automatically convert all ingredients into same category?

sc-arecrow replied:
Yes, an invariant of an `IngredientCategory` is that all ingredients inside have the same `ingredientCategoryId` as the `id` of the category.

chrystalquek commented on ChopChop/Models/Ingredients/Quantity.swift:
```diff
@@ -0,0 +1,258 @@
+/**
```
This file is inevitably quite duplicated but I think we somehow need to explain the duplicated code in the DG

sc-arecrow replied:
Yes I agree, I'll see if I can reduce the duplication by using `QuantityType`.

chrystalquek commented on ChopChop/Models/Ingredients/Quantity.swift:
```diff
@@ -0,0 +1,258 @@
+
+// MARK: - CustomStringConvertible
+extension Quantity: CustomStringConvertible {
+    var description: String {
```
does this part display the quantity magnitude sensibly ie 50ml vs 0.05 litre

sc-arecrow replied:
Thanks for the suggestion, I'll add it in!

chrystalquek commented on ChopChop/Models/Ingredients/Ingredient.swift:
```diff
@@ -0,0 +1,271 @@
+    /**
+     Removes all batches with expiry dates earlier than the current date.
+     */
+    func removeExpiredBatches() {
```
interesting, are we gonna call this method periodically from somewhere?

sc-arecrow replied:
This seemed like a useful function to have (maybe there will be a button that users can press to rid an ingredient of all its expired batches)

chrystalquek commented on ChopChop/Models/Ingredients/Ingredient.swift:
```diff
@@ -0,0 +1,271 @@
+    }
+}
+
+enum IngredientError: Error {
```
i'll follow your way of putting the enum errors in the file itself

sc-arecrow replied:
Should we discuss if having it in a separate file is better? I'm open to both ways

chrystalquek replied:
I think we can leave it like this for now, I don't think one way is better than the other

seowalex commented on ChopChop/Models/Ingredients/Ingredient.swift:
```diff
@@ -0,0 +1,271 @@
+        self.batches = batches
+    }
+
+    required init(row: Row) {
+        id = row["id"]
+        ingredientCategoryId = row["ingredientCategoryId"]
+        quantityType = .count
+        name = row["name"]
+        batches = row.prefetchedRows["ingredientSets"]?.map {
+            let record = IngredientBatchRecord(row: $0)
+
+            return IngredientBatch(quantity: record.quantity, expiryDate: record.expiryDate)
+        } ?? []
+    }
```
Should this be in an extension to keep the database logic separate?

sc-arecrow replied:
I believe the initialiser for `FetchableRecord` has to be `required`, and `required` initialisers for classes cannot be kept in extensions. I think this wasn't an issue previously when `Ingredient` was a struct.

seowalex commented on ChopChop/Models/Ingredients/Ingredient.swift:
```diff
@@ -0,0 +1,271 @@
+     - Throws:
+        - `IngredientError.emptyName`: if the given name is empty.
+     */
+    func rename(_ newName: String) throws {
+        let trimmedName = newName.trimmingCharacters(in: .whitespacesAndNewlines)
+
+        guard !trimmedName.isEmpty else {
+            throw IngredientError.emptyName
+        }
+
+        self.name = newName
+    }
```
Can errors be thrown from property observers (`willSet` and `didSet`)? If they can maybe we should use that instead.

sc-arecrow replied:
I think [it's a Swift limitation](https://stackoverflow.com/questions/34797704/throwing-errors-inside-observer-methods-in-swift); you can't do `willSet throws {`.

seowalex commented on ChopChop/Models/Ingredients/Ingredient.swift:
```diff
@@ -0,0 +1,271 @@
+            throw QuantityError.differentTypes
+        }
+
+        let existingBatches = batches.filter { batch in
+            batch.expiryDate == expiryDate
+        }
+
+        if let existingBatch = existingBatches.first {
```
Can't you just use [`first(where:)`](https://developer.apple.com/documentation/swift/array/1848165-first)?

sc-arecrow replied:
Thanks for the suggestion!

seowalex commented on ChopChopTests/Models/Ingredients/QuantityTests.swift:
```diff
@@ -0,0 +1,287 @@
+    func testLessThan_sameQuantityType_success() {
+        let leftCount: Quantity = .count(1)
+        let rightCount: Quantity = .count(2)
+        XCTAssertTrue(try leftCount < rightCount)
+        XCTAssertFalse(try rightCount < leftCount)
```
Any reason why relevant methods like `XCTAssertLessThan` are not used?

sc-arecrow replied:
Thanks for the suggestion!

seowalex commented on ChopChopTests/Models/Ingredients/QuantityTests.swift:
```diff
@@ -0,0 +1,287 @@
+        let leftCount: Quantity = .count(1)
+        let rightCount: Quantity = .count(1)
+        let differentCount: Quantity = .count(2)
+        XCTAssertTrue(leftCount == rightCount)
+        XCTAssertFalse(leftCount == differentCount)
```
Same here but for `XCTAssertEqual`.

sc-arecrow replied:
Thanks for the suggestion!

seowalex commented on ChopChopTests/Models/Ingredients/IngredientBatchTests.swift:
```diff
@@ -0,0 +1,140 @@
+            quantity: IngredientBatchTests.existingQuantity,
+            expiryDate: laterDate)
+
+        XCTAssertTrue(batch < laterBatch)
```
Any reason why `XCTAssertLessThan` isn't used?

sc-arecrow replied:
Thanks for the suggestion!

seowalex commented on ChopChop/Models/Ingredients/IngredientReference.swift:
```diff
@@ -0,0 +1,24 @@
+    let name: String
+    private(set) var quantity: Quantity
+
+    init(name: String, quantity: Quantity) {
+        self.name = name
+        self.quantity = quantity
+    }
```
Doesn't seem to have the same checks `Ingredient` has, which seems odd. If you create an `IngredientReference` with an empty name, it's not going to match any ingredient.

sc-arecrow replied:
I agree that it should have the same checks, I'll add it in

seowalex commented on ChopChop/Models/Ingredients/IngredientCategory.swift:
```diff
@@ -0,0 +1,59 @@
+     If the category contains an existing ingredient with the same name and quantity type,
+     combines the given ingredient instead.
+     */
+    func add(_ addedIngredient: Ingredient) {
+        let sameIngredient = ingredients.first { ingredient in
+            ingredient == addedIngredient
+        }
+
+        guard let existingIngredient = sameIngredient else {
+            ingredients.append(addedIngredient)
+            addedIngredient.ingredientCategoryId = id
+            return
+        }
+
+        do {
+            try existingIngredient.combine(with: addedIngredient)
+        } catch {
+            return
+        }
+    }
```
How will this be synchronised with the database?

sc-arecrow replied:
It currently does not, I'll work on it

seowalex commented on ChopChop/Models/Ingredients/Quantity.swift:
```diff
@@ -0,0 +1,258 @@
+    case mass(Double)
+    case volume(Double)
+
+    var type: QuantityType {
+        switch self {
+        case .count:
+            return .count
+        case .mass:
+            return .mass
+        case .volume:
+            return .volume
+        }
+    }
+
+    var value: Double {
+        get {
+            switch self {
+            case .count(let value):
+                return value
+            case .mass(let value):
+                return value
+            case .volume(let value):
+                return value
+            }
+        }
+        set {
+            switch self {
+            case .count:
+                self = .count(newValue)
+            case .mass:
+                self = .mass(newValue)
+            case .volume:
+                self = .volume(newValue)
+            }
+        }
+    }
```
Given that you have to do this, might it not be a better idea to have a `Quantity` class instead, with a `quantityType` and `value`? My quantity enum wasn't actually meant to be used for the actual quantities, I was intending to map from whatever `Quantity` object you had.

sc-arecrow replied:
I think this idea might solve a lot of the duplication problems, I'll try redesigning it

sc-arecrow replied:
I'm going ahead with the redesign for now but a small problem I just thought of is how to deal with a new `Quantity` type whose `value` is not a `Double`

seowalex replied:
I imagine since the value has to be a number, a `Double` probably covers all possible cases.

seowalex commented on ChopChop/Models/Ingredients/Quantity.swift:
```diff
@@ -0,0 +1,258 @@
+        switch (left, right) {
+        case let (.count(leftValue), .count(rightValue)):
+            let sum = leftValue + rightValue
+            guard sum >= 0 else {
+                throw QuantityError.negativeQuantity
+            }
```
Is it better to enforce that quantities cannot be negative?

sc-arecrow replied:
Will be done when `Quantity` is redesigned to be a class

seowalex commented on ChopChop/Models/Ingredients/Ingredient.swift:
```diff
@@ -0,0 +1,271 @@
+                continue
+            }
+
+            do {
+                try subtractedQuantity -= batch.quantity
+                usedBatches.append(batch)
+            } catch QuantityError.negativeQuantity {
+                try batch.subtract(subtractedQuantity)
+                subtractedQuantity.value = 0
+                break
+            }
```
Seems like a repetition of the above logic for batches with no expiry date.

sc-arecrow replied:
I'll see if I can refactor it out, thanks!

seowalex reviewed (commented):
Just a question for `QuantityType`: What's the benefit of using that over checking the batches to see if they have matching types?

sc-arecrow commented:
> Just a question for `QuantityType`: What's the benefit of using that over checking the batches to see if they have matching types?

If you're asking why `Ingredient` has a `QuantityType` field, it's because you can't check the batches inside the ingredient if there isn't any batches inside.

sc-arecrow committed:
Refactor code and add tests

sc-arecrow committed:
Redesign Quantity to a class

seowalex commented on ChopChop/Models/Ingredients/IngredientCategory.swift:
```diff
@@ -0,0 +1,87 @@
+        name = row["name"]
+        ingredients = []
+
+        if let categoryId = id {
+            let storageManager = StorageManager()
+            storageManager.ingredientsFilteredByCategoryOrderedByNamePublisher(ids: [categoryId])
+                .sink(receiveCompletion: { _ in }, receiveValue: { [weak self] infos in
+                    let ingredientIds = infos.compactMap { $0.id }
+                    let ingredients = ingredientIds.compactMap { try? storageManager.fetchIngredient(id: $0) }
+                    self?.ingredients = ingredients
+                })
+        }
```
This is actually not an intended use of the publisher. The publishers are intended to be used in the view models to update the list of objects to be rendered. If you want I can either

1. Provide a method to fetch ingredients given a categoryId
2. Load the ingredients together with the category to begin with so you can avoid doing this

sc-arecrow replied:
I think 1. would be good

seowalex commented:
You don't seem to have an easy way to add different quantities of ingredients (e.g. tablespoons, teaspoons, cups, grams, millilitres), which might be useful in the future (e.g. after parsing the list of ingredients of a recipe).

sc-arecrow commented:
> You don't seem to have an easy way to add different quantities of ingredients (e.g. tablespoons, teaspoons, cups, grams, millilitres), which might be useful in the future (e.g. after parsing the list of ingredients of a recipe).

I think @shadowezz is working on conversions

sc-arecrow committed:
Merge branch 'main' of https://github.com/cs3217-chopchop/ChopChop into main

sc-arecrow committed:
Merge branch 'main' into sc-arecrow/ingredients

sc-arecrow committed:
Redesign IngredientCategory

Removed tests for now because unit testing IngredientCategory
would require some kind of StorageManager fake

sc-arecrow committed:
Add name restriction to ingredient categories

sc-arecrow committed:
Rename IngredientReference to RecipeIngredient
